design/cluster-scope-resource-filter.md

```diff
@@ -0,0 +1,219 @@
+# Proposal to separate resource filter into cluster scope and namespace scope.
+
+## Abstract
+The current filter (IncludedResources/ExcludedResources + IncludeClusterResources flag) is not enough for some special cases, e.g. all namespaced resources + some kind of cluster resource and all namespaced resources + cluster resource excludes.
+Propose to separate `IncludedResources` into two cluster scope and namespace scope parameters, and same change for `ExcludedResources`.
+
+## Background
+There are two sets of resource filters for Velero: `IncludedNamespaces/ExcludedNamespaces` and `IncludedResources/ExcludedResources`. 
+`IncludedNamespaces` means only including the resource types specified in the parameter.
+The k8s resources are separated into cluster scope and namespaced scope.
+As a result, it's hard to include all resources in one group and only including specified resource in other group.
+
+## Goals
+- Make the existing resource filters supporting more scenarios without conflict.
+
+## Non Goals
+- Enrich the resource filtering rules.
+
+
+## High-Level Design
+Two new parameters are added into command `velero backup create`: `--include-cluster-scope-resources` and `--exclude-cluster-scope-resources`. 
+They are used to filter cluster scope resources included or excluded in backup.
+`--include-cluster-scope-resources` valid value includes `*` and comma separated string.
+`--exclude-cluster-scope-resources` valid value is comma separated string.
+
+### Using scenarios:
+#### some namespaced resources + no additional cluster resources
+``` bash
+velero backup create <backup-name>
+    --include-namespaces=default,kube-system
+```
+
+#### some namespaced resources + some additional cluster resources
+``` bash
+velero backup create <backup-name>
+    --include-namespaces=default,kube-system
+    --include-cluster-scope-resources=StorageClass
+```
+
+#### some namespaced resources + all cluster resources
+``` bash
+velero backup create <backup-name>
+    --include-namespaces=default,kube-system
+    --include-cluster-scope-resources=*
+```
+
+or
+
+``` bash
+velero backup create <backup-name>
+    --include-namespaces=default,kube-system
+    --include-cluster-resources=true
+```
```

Comment:
These two have different meanings, so I don't think those two are equivalent. IncludeClusterResources is a bool pointer with 3 possible values. `true` means all cluster resources are included (still subject to other filters such as label selector and -- with the current codebase -- include/exclude-resources). `false` means all are excluded, regardless of the other selection criteria. And `nil` means "only consider cluster resources that are directly relevant to namespaced resources being included -- specifically, this means PVs when bound to included PVCs, CRDs when backing up associated CRs, and other cluster resources as pulled in by backup plugins. I think if we're adding the new Include/ExcludeClusterScopeResources fields, we still need to factor in include-cluster-resources. For example, you could set it to true and then use the new fields to fine tune which resources you want. That way you can still set it to nil and only bring in relevant ones (i.e. so you don't pull in every PV), and still exclude things like CRDs, etc.

Reply:
What about moving the same logic in `include-cluster-resources` to `IncludeClusterScopeResources`?

Reply:
Modified.